File: src/Python/264.UglyNumberII.py

class Solution:
    def nthUglyNumber(self, n):
        dp = [0] * n
        dp[0] = 1
        p2 = 0
        p3 = 0
        p5 = 0
        for i in range(1, n):
            dp[i] = min(2 * dp[p2], 3 * dp[p3], 5 * dp[p5])
            if dp[i] >= 2 * dp[p2]:
                p2 += 1
            if dp[i] >= 3 * dp[p3]:
                p3 += 1
            if dp[i] >= 5 * dp[p5]:
                p5 += 1
        return dp[-1]

    
if __name__ == "__main__":
    solution = Solution()
    print(solution.nthUglyNumber(10))

File: src/CSharp/264.UglyNumberII.cs

using System;
using System.Linq;
using System.Text;
using System.Collections.Generic;

namespace PeefyLeetCode.UglyNumberII
{

    public class Solution
    {
        public int NthUglyNumber(int n)
        {
            int[] dp = new int[n];
            dp[0] = 1;
            int p2 = 0, p3 = 0, p5 = 0;
            for (int i = 1; i < n; i++)
            {
                dp[i] = Math.Min(dp[p2] * 2, dp[p3] * 3);
                dp[i] = Math.Min(dp[i], dp[p5] * 5);
                if (dp[i] >= dp[p2] * 2)
                    p2 += 1;
                if (dp[i] >= dp[p3] * 3)
                    p3 += 1;
                if (dp[i] >= dp[p5] * 5)
                    p5 += 1;
            }
            return dp[n - 1];
        }
    }
}